### PR TITLE
DT: Force skip levels 

### PR DESCRIPTION
**Pull request by adamkecskes**
### New logic for level locking/skipping:
 Old logic: Marking one level from which all levels will be locked
 New logic: Each level can be marked as locked and/or optional individually

Marking logic: `locked` flag can be added with a date until which the lock will be effective.
 
 How we store the data?
 
 ```javascript
 classroom: {
   attributes: {
     studentLockMap: {
       lockedLevels: {
         [level1Original]: true // level1 is locked
         [level2Original]: Date // level2 is locked until Date
       },
       {
         optionalLevels: {
           [level1Original]: true // level1 is optional !! because level1 is also locked, this level will be shown as skipped (i.e. unaccessible) !! 
           [level3Original]: true // level3 is optional. It'll be accessible, student will be able to play it, but completion is not a prerequisite for accessing further levels 
         }
       }
     }
   }
 }
 ```

### **Teacher view:**
 On the Teacher Dashboard you'll be able to select multiple levels (in the table header) and students (left side of the table). 
 If you select them, two buttons will become active(shown):
  - Module Lock button - by clicking it each level in the module will be selected, and you'll be able to lock/skip/make-optional them 
  - Lock or Skip Levels button - by clicking it, you'll be able to lcok/skipt/optional the selected levels for the selected students
  
dt-skip

### **Student view:**
  Nothing much changed, the student's will be only able to play unskipped levels, and also will be able to choose if they'd like to play the optional ones.

Cursor_and_Students

### CODE REVIEW

**Comment by smallst** [comment]
the overview looks good to me. i can test it by tomorrow.

File: app/views/play/CampaignView.coco.coffee [hunk 1]
```diff
@@ -1502,7 +1502,8 @@ module.exports = class CampaignView extends RootView
 
       level.noFlag = !level.next
 
-      if level.slug == @courseInstance.get('startLockedLevel') # lock level begin from startLockedLevel
+#      if level.slug == @courseInstance.get('startLockedLevel') # lock level begin from startLockedLevel
```
Comment (smallst) [inline app/views/play/CampaignView.coco.coffee]: remove

File: app/views/play/CampaignView.coco.coffee [hunk 2]
```diff
@@ -1502,7 +1502,8 @@ module.exports = class CampaignView extends RootView
...
+      if this.courseInstance.get('lockedLevels').includes level.slug
```
Comment (smallst) [inline app/views/play/CampaignView.coco.coffee]: `...get('lockedLevels')?.includes`

File: ozaria/site/store/BaseSingleClass.js [hunk 3]
```diff
@@ -174,22 +217,35 @@ export default {
...
+
+          const levelsToHandle = levels.filter((level) => {
+            if (value) {
+              return !ClassroomLib.isModifierActiveForStudent(clonedClass, _id, currentCourseId, level, modifier, date)
+            } else {
+              return ClassroomLib.isModifierActiveForStudent(clonedClass, _id, currentCourseId, level, modifier, date)
+            }
+          })
+
+          console.log('modifier', modifier, 'levelsToHandle', levelsToHandle)
```
Comment (smallst) [inline ozaria/site/store/BaseSingleClass.js]: remove

File: app/templates/courses/teacher-class-view.pug [hunk 4]
```diff
@@ -217,8 +217,9 @@ mixin courseOverview(course)
         - var courseInstance = view.getSelectedCourseInstance();
         each level, index in levels
           - if (level.get('assessment')) continue;
-          - if (courseInstance && level.get('slug') === courseInstance.get('startLockedLevel')) lockedFromHere = true;
+          - if (courseInstance && level.get('slug') === courseInstance.get('lockedLevels')) levelLocked = true;
           - var progress = state.get('progressData').get({ classroom: view.classroom, course: course, level: level })
+          - console.log(level.get('slug'), progress)
```
Comment (smallst) [inline app/templates/courses/teacher-class-view.pug]: remove

**Review by smallst — CHANGES_REQUESTED** [review]
LGTM. left some comments but already push the fix.
BTW can we update such UI to say `skipped` instead of `locked`? 
image

**Comment by adamkecskes** [comment]
@smallst UI now shows Skipped/Optional:
levelstatus

**Pull request merged**